packages/ActionBar/stories/ControlledActionBar/MyFilter.js

```diff
@@ -0,0 +1,120 @@
+import React from "react";
+import nanoid from "nanoid";
+import { Filter } from "../../src";
+
+const rulesByType = {
+  ...Filter.defaultRulesByType,
+  SINGLE_SELECT: [Filter.rules.IS, Filter.rules.IS_NOT, Filter.rules.IS_EMPTY, Filter.rules.IS_NOT_EMPTY],
+};
+
+const getDefaultFilter = () => {
+  return {
+    columnId: "goals",
+    rule: "LESS_THAN",
+    value: "0",
+    id: nanoid(),
+  };
+};
+
+const initialValueByType = {
+  BOOLEAN: true,
+  NUMBER: "",
+  TEXT: "",
+  DATE: "",
+  SINGLE_SELECT: null,
+};
+
+const levelSelect = () => (
+  <select>
+    <option value="low">low</option>
+    <option value="mid">mid</option>
+    <option value="high">high</option>
+  </select>
+);
+
+const getLevelFilter = () => {
+  return {
+    columnId: "level",
+    rule: "IS",
+    value: null,
+    id: nanoid(),
+    renderValueField: levelSelect,
+  };
+};
+
+export default function MyFilter({ columns }) {
+  const [filters, setFilters] = React.useState([]);
+  const [operator, setOperator] = React.useState("AND");
+
+  function handleFilterChange({ id: filterId, rule, value, columnId }) {
+    let change;
+
+    if (columnId) {
+      const columnType = columns.find(column => column.id === columnId).type;
+      if (columnType === "SINGLE_SELECT") {
+        change = getLevelFilter();
+      } else {
+        change = {
+          columnId,
+          rule: Filter.defaultRulesByType[columnType][0],
+          value: initialValueByType[columnType],
+          renderValueField: null,
+        };
+      }
+    } else if (rule) {
+      change = {
+        rule,
+        value: "",
+      };
+    } else {
+      change = {
+        value,
+      };
+    }
+    setFilters(prevFilters =>
+      prevFilters.map(filterItem => (filterItem.id === filterId ? { ...filterItem, ...change } : filterItem))
+    );
+  }
+
+  const memorizedHandleChange = React.useCallback(handleFilterChange, [columns, setFilters]);
+
+  function handleAddFilter() {
+    setFilters(prevFilters => [...prevFilters, getDefaultFilter()]);
+  }
+
+  const memorizedAddFilter = React.useCallback(handleAddFilter, [setFilters]);
+
+  function handleDeleteFilter(deletedFilterId) {
+    setFilters(prevFilters => [...prevFilters].filter(filter => filter.id !== deletedFilterId));
```

Comment:
There is no need to do [...prevFields], because filter will return a new array anyway.